### ISSUE TITLE
Transformations of kernels

### ISSUE THREAD

**Issue by willtebbutt** [theogf/KernelFunctions.jl#17]
Currently, kernels have a field `transform` e.g. [here](https://github.com/theogf/KernelFunctions.jl/blob/460ed8685018a35bac2bceaa2401639668a41f6e/src/kernels/exponential.jl#L12).

This is unnecessary, and creates additional code to maintain. I would argue that a better approach is to orthogonalise the design by removing the `transform` field from the kernels and creating a combinator kernel `Transform` that wraps the kernel. i.e.

```julia
struct TransformedKernel{Tt<:Transform, Tk<:Kernel} <: Kernel
    t::Tt
    k::Tk
end

function kernel_matrix(t::TransformedKernel, x::AbstractVector{<:Real})
    return kernel_matrix(t.k, apply(t.t, x))
end
```

This has the distinct advantage that you can transform the inputs to any kernel. For example, the [KernelSum](https://github.com/theogf/KernelFunctions.jl/blob/460ed8685018a35bac2bceaa2401639668a41f6e/src/kernels/kernelsum.jl#L14) currently doesn't have a `transform` field, nor should it. This functionality would be obtained for free via this design.

Stheno currently does something similar to this with [stretch](https://github.com/willtebbutt/Stheno.jl/blob/61236affc381d5cba16fab67b9a0d93226e8995d/src/gp/kernel.jl#L438), but I'm planning to move more closely towards the above design over time.

**Comment by theogf** [comment]
I think here we are mostly dealing with a UI problem. I still maintain the idea that the `Kernel` type should contain `transform`. I like to be able to directly see what kind of kernel is used without having to check subparameters. With the solution you propose literally everything would be a `TransformedKernel`... 
> This has the distinct advantage that you can transform the inputs to any kernel. For example, the KernelSum currently doesn't have a transform field, nor should it. This functionality would be obtained for free via this design.
About this how about (taking #22 into account)
```
abstract type Kernel end
abstract type SingleKernel{Tr<:Transform} <: Kernel end
abstract type CompositeKernel <: Kernel end
struct KernelSum <: CompositeKernel
...
end
```
I don't like `SingleKernel` but it's just for the example

**Comment by willtebbutt** [comment]
I partly agree it's a UI problem, but I'm more concerned about the maintainability of the internals and composability.

Maintability: having the transformation being part of the type adds extra code to maintain to every single kernel. This is as opposed to having the transformations being completely separate from the "base" kernels.

Composability: having the transform outside of the kernel is also composable. Consider, for example, the current implementation of `KernelSum`. It doesn't have a transform field, meaning that it's not currently possible to jointly transform the inputs to all of the kernels. While I'm sure this can be fixed, the point is that it would simply never be an issue if the transformation is outside of the kernel.

UI: My interpretation of this problem is that it's ultimately a problem in terms of how kernels are printed to the REPL, and I agree that the default printing of kernels implemented in the way that I'm suggesting is annoying (please correct me if this isn't your concern). My solution, therefore, would be to take more care with how kernels are printed to minimise the problem of parsing a kernel. It's something that I think we should consider addressing anyway, because any nested structure involving compositions of kernels is going to have this issue, regardless of whether or not we have the transformation inside or out.

Sorry, it's not clear to me what you mean by `SingleKernel` here. Is this meant to wrap a "base" kernel?

**Comment by devmotion** [comment]
I implemented yesterday a simple tensor product kernel type and also noticed that it is quite strange and unnatural that I somehow had to provide a transform and work around some hard-coded accesses to `kernel.transform` in the current implementation.

**Comment by theogf** [comment]
Would you share the code about this? I am curious to see where the transform field approach fails

**Comment by theogf** [comment]
> Sorry, it's not clear to me what you mean by SingleKernel here. Is this meant to wrap a "base" kernel?

I meant a `BaseKernel` indeed

> Having the transformation being part of the type adds extra code to maintain to every single kernel. This is as opposed to having the transformations being completely separate from the "base" kernels.

So you convinced me that `BaseKernel` (i.e `SqExponential`, `LinearKernel` etc), should have `transform` as a field and I agree we should change this.

>  I would argue that a better approach is to orthogonalise the design by removing the transform field from the kernels and creating a combinator kernel Transform that wraps the kernel.

However I am not convinced that the solution is to create a `TransformedKernel` like in `Stheno` :
Not everyone will want to have crazy transformations and I think it's important to let people create simple kernels like `SqExponentialKernel(3.0)` where the `ScaleTransform` is implicit. 
With `TransformedKernel` the constructor would just create a different object (you pointed out this problem for `MaternKernel` some months ago for specific `nu`).

**Comment by willtebbutt** [comment]
> Not everyone will want to have crazy transformations and I think it's important to let people create simple kernels like SqExponentialKernel(3.0) where the ScaleTransform is implicit.

I agree with you that it's important to have a sensible interface for constructing standard kernels, but this doesn't prescribe a particular implementation of the internals. If you have two methods to construct a squared exponential kernel:
```julia
k = sq_exponential_kernel()
k = sq_exponential_kernel(3.0)
```
the first returning an object of type `SqExponentialKernel`, and the second `TransformedKernel{SqExponentialKernel, TransformType}` (or whatever particular names we settle on), the user never needs to know that the second method returns a `TransformedKernel` type if they don't care to look at the internals. This satisfies the requirement that you can get an object that represents a squared exponential kernel with a particular length scale, but it uses an internal representation that is easier to maintain and test.

> With TransformedKernel the constructor would just create a different object (you pointed out this problem for MaternKernel some months ago for specific nu).

[This](https://github.com/theogf/KernelFunctions.jl/issues/7) is different. In this case the type of the return type of the `Matern` constructor depended upon run-time info (a specific numeric value) causing type instabilities. This would not be the case under the proposal here -- if you provide a length scale then a transformed thing is returned (even if you provide the numeric value `1`), if you don't provide a length scale you get the base thing.

> So you convinced me that BaseKernel (i.e SqExponential, LinearKernel etc), should have transform as a field and I agree we should change this.

This isn't really what I was trying to get at. I want to be able to transform each base kernel, but I don't want us to have to maintain a `transform` field for each of the base kernels. Phrased differently: I want us to have to implement the notion of length-scale once, and have it be applicable to any existing base kernel.

I completely agree with your goal of making standard kernels easy to construct, but I want to implement it in a way that maximises code re-use and minimises the opportunities for inconsistencies between different implementations of e.g. length scales.

**Comment by theogf** [comment]
>This satisfies the requirement that you can get an object that represents a squared exponential kernel with a particular length scale, but it uses an internal representation that is easier to maintain and test.

This sounds like a reasonable option, I like it! Now I guess we should agree on the formulation of these external constructors : 
- Should we keep the struct constructor style and keep uppercase for the first letters but use different names (maybe remove the `Kernel` suffix, I know that would make you happy @willtebbutt :smile: )
- Should we use only lower case and keep the same naming ?
- Others...

> So you convinced me that BaseKernel (i.e SqExponential, LinearKernel etc), should have transform as a field and I agree we should change this.

Ups sorry I meant "should not" :smile:

**Comment by theogf** [comment]
Solved with #32
